### ISSUE TITLE
 SecurityPolicy -> HTTPRoute  Auth mismatch / broken Envoy config

### ISSUE THREAD

**Issue by apjoseph** [envoyproxy/gateway#2326]
*Description*:
I've continued testing my very basic ouath2 app +jwt api SecurityFilter use case and have encountered what seems to be substantially missing/broken functionality  when dealing with different security filters on different routes. on the same or different backendRef services. 

I have two deployments 

- a react app using host `dev.example.com` where authentication should be handled by ouath2
- and a node app using host `api.dev.example.com` which should be secured by JWTs 

I discovered that it is not currently not to call the API from the frontend at `api.dev.example.com` with Envoy Gateway because Envoy itself currently does not allow setting the domain of a cookie. I would need domain=dev.example.com in order for CORS to allow cookies for `dev.example.com` to be passed to `api.dev.example.com`

To work around this limitation, I decided to try send requests to `dev.example.com/api ` to my API service and use an HTTPRoute to redirect anything at /api to the node.js backend.

From my understanding of the Gateway, I used the following route/ security configuration

### Routes
- /api

### Upstream Service
Backend Node API

### Configuration
```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: {{.Values.web.name}}-web-api
  labels:
    {{- include " example.labels" . | nindent 4 }}
    {{- include " example.web.labels" . | nindent 4 }}
  annotations:
    external-dns.alpha.kubernetes.io/hostname: {{ include " example.baseDomainName" . }}
spec:
  parentRefs:
    - name: {{.Values.web.name}}-web
      sectionName: {{.Values.web.name}}-web
  rules:
    - backendRefs:
        - group: ""
          kind: Service
          name: {{.Values.web.name}}-web-api
          namespace: {{.Release.Namespace}}
          port: 8080
      matches:
        - path:
            type: PathPrefix
            value: /api


---

apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: {{.Values.web.name}}-web-api
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: {{.Values.web.name}}-web-api

  jwt:
    providers:
      - name: auth0
        extractFrom:
          cookies:
            - BearerToken
        remoteJWKS:
          uri: https://{{include " example.auth.domain" .}}/.well-known/jwks.json
```

## OAuth2
###Routes
- / oauth2/callback (hardcoded Envoy callback route)
- /login
- /signout (hardcoded Envoy logout route) 

### Upstream Service
React App - though in this case Envoy is  

### Configuration

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: {{.Values.web.name}}-web-oauth2
  labels:
    {{- include " example.labels" . | nindent 4 }}
    {{- include " example.web.labels" . | nindent 4 }}
  annotations:
    external-dns.alpha.kubernetes.io/hostname: {{ include " example.baseDomainName" . }}
spec:
  parentRefs:
    - name: {{.Values.web.name}}-web
      sectionName: {{.Values.web.name}}-web
  rules:
    - backendRefs:
        - group: ""
          kind: Service
          name: {{.Values.web.name}}-web
          namespace: {{.Release.Namespace}}
          port: 8080
      matches:
        - path:
            type: PathPrefix
            value: /oauth2/callback
        - path:
            type: PathPrefix
            value: /login
        - path:
            type: PathPrefix
            value: /signout

---

apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: {{.Values.web.name}}-web-oauth2
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: {{.Values.web.name}}-web-oauth2
  oidc:
    clientID: {{include " example.web.appClientID" .}}
    clientSecret:
      name: {{.Values.web.name}}-web-client-secret
    provider:
      issuer: https://{{include "example.auth.domain" .}}
```

### Routes
- /* (all other routes) 
### Upstream Service
React App

### Configuration
```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: {{.Values.web.name}}-web
  labels:
    {{- include " example.labels" . | nindent 4 }}
    {{- include " example.web.labels" . | nindent 4 }}
  annotations:
    external-dns.alpha.kubernetes.io/hostname: {{ include " example.baseDomainName" . }}
spec:
  parentRefs:
    - name: {{.Values.web.name}}-web
      sectionName: {{.Values.web.name}}-web
  rules:
    - backendRefs:
        - group: ""
          kind: Service
          name: {{.Values.web.name}}-web
          namespace: {{.Release.Namespace}}
          port: 8080
      matches:
        - path:
            type: PathPrefix
            value: /
```

I tried several variations. on the above. Each time I received 500/503 responses with envoy deployment logs showing messages like:

```shell
OAuth2 filter: unknown cluster 'auth_provider_com_443' in config. Please specify which cluster to direct OAuth requests to.
[2023-12-20 00:03:10.249][1][error][filter] [source/extensions/filters/http/common/jwks_fetcher.cc:51] fetch: fetch pubkey [uri = https://auth.provider.com/.well-known/jwks.json] failed: [cluster = auth_provider_com_443] is not configured
```

If I used the same service in two different routes, the gateway translator seems to merge them and then put everything under oauth2 and ignore the more route specific settings. 

The TLDR; is that every possible route/security policy configuration I could think of attempting failed by either by breaking the envoy configuration or applying a variety of wrong combinations of auth/route. Again this is just a basic oauth2 app using an api with a Bearer token cookie, nothing too fancy.

**Comment by Xunzhuo** [comment]
Thanks @apjoseph, recently we have recieved a lot of valuable feedbacks and reports from you.

cc @zhaohuabing PTAL

**Comment by zhaohuabing** [comment]
@apjoseph  why you defined an HTTPRoute for oauth2 callback and logout path? They're not needed.

https://gateway.envoyproxy.io/latest/user/oidc/

**Comment by apjoseph** [comment]
@zhaohuabing  my goal there was not so much to define a route, but to restrict the oauth2 filter to the specific routes that envoy's oauth2 filter handles.

If I put the oidc filter at / then it requires credentials to access every page on the app.

The route /oauth2/callback is handled by envoy and does the PKCE code exchange and adds the cookies before sending the user back to the page they came from.

/login should never really hit the backend since the oidc filter should redirect to the odic provider's authorization endpoint if they aren't already logged in (Though for completeness I should have added that as a redirect route in case they are already logged in)

So basically I was attempting to have envoy do PKCE flow for my app without requiring authentication for every page in the app.

Is there another way I can achieve that?

**Comment by apjoseph** [comment]
@zhaohuabing The oauth 2 filter defines a `pass_through_matcher` that allows routes to be excluded from the oauth2filter

So if I did put the oauth2 SecurityFilter at / then I think the oidc configuration for SecurityFilter needs an option like 

`passThrough: bool` that adds the route to the pass_through_matcher.

Same for JWT. Otherwise you can't really configure security on a route by route basis.

**Comment by zhaohuabing** [comment]
> If I put the oidc filter at / then it requires credentials to access every page on the app.

> The route /oauth2/callback is handled by envoy and does the PKCE code exchange and adds the cookies before sending the user back to the page they came from.

`oauth2/callback` is the default callback path used by Envoy Gateway. The request to this url is handled by Envoy, and will never be forwarded to the backend. Therefore, you don't need to define a route for `oauth2/callback`.

> /login should never really hit the backend since the oidc filter should redirect to the odic provider's authorization endpoint if they aren't already logged in (Though for completeness I should have added that as a redirect route in case they are already logged in)

You don't necessarily need a `login` url for OIDC, but you can define one and point to a page in your backend.

> So basically I was attempting to have envoy do PKCE flow for my app without requiring authentication for every page in the app.

The above should work if you follow the simple steps in this doc https://gateway.envoyproxy.io/latest/user/oidc/

**Comment by zhaohuabing** [comment]
@apjoseph any update on this issue?

**Comment by apjoseph** [comment]
@zhaohuabing Yes. My initial post was a bit all over the place, so to narrow this down and reduce confusion, I'll focus on the main issue that I was able to isolate after further testing: 

The oauth2/callback endpoint, which envoy manages, is only activated when the SecurityPolicy defining the oidc section is applied to an HttpRoute with a backend without a path match OR a path match of /

So If I only apply the oidc SecurityPolicy to an HttpRoute with say a path match of /secret-page then the oauth2 filter will intercept the request and redirect to the defined authorization endpoint BUT it won't handle the PKCE exchange at /oauth2/callback when the authorization service redirects. This results in the callback request being passed directly to the backend.

Let me know if this provides enough clarification. I'm a bit tied up with a project at the moment, but I can provide more  details and examples if it would help.

**Comment by zhaohuabing** [comment]
@apjoseph I think I got what you mean. Could you please paste an example of HTTPRoutes and the corresponding `SecuritPolicy` of that situation?

**Comment by zhaohuabing** [comment]
@apjoseph I think we can solve this by exposing callback url to OIDC setting. So users can define different callback url for different route. For example, `/secret-page/oauth2/callback` can be used for the route match of `/secret-page` 

HTTPRoute
```
kind: HTTPRoute
metadata:
  name: backend
spec:
  hostnames:
  - www.example.com
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: eg
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /
    backendRefs:
    - group: ""
      kind: Service
      name: backend
      port: 3000
  - matches:
    - path:
        type: PathPrefix
        value: /secret-page
    backendRefs:
    - group: ""
      kind: Service
      name: backend
      port: 3000
```
SecurityPolicy
```yaml
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: oidc-example
spec:
  oidc:
    redirect_uri: /secret-page/oauth2/callback
    clientID: 250344188863-uve9br9mtrj05j2tfn8ls7n3fhemg3sr.apps.googleusercontent.com
    clientSecret:
      group: ""
      kind: Secret
      name: my-app-client-secret
    provider:
      issuer: https://accounts.google.com
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: backend
  ```

```

**Comment by apjoseph** [comment]
@zhaohuabing What about creating an option like `passthrough: true` in the oidc config and utilizing the passthrough matcher?

[(pass_through_matcher](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/oauth2/v3/oauth.proto#envoy-v3-api-msg-extensions-filters-http-oauth2-v3-oauth2) 
> (repeated [config.route.v3.HeaderMatcher](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-msg-config-route-v3-headermatcher)) Any request that matches any of the provided matchers will be passed through without OAuth validation.

My reasoning is that oauth providers typically require that you manually specify the callback url. So if you create /another-secret-page you'd need to manually add that to your provider configuration (as they don't typically support regular expressions) which could result in large amounts of boilerplate. 

It's much more convenient to have opt-out functionality for security filters when they apply to many routes. 

A more comprehensive solution would be a CRD like `SecurityConfiguration` where you can define something like 

```
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityConfiguration
metadata:
  name: oidc-config-default
spec:
  oidc:
    redirect_uri: /secret-page/oauth2/callback
    clientID: 250344188863-uve9br9mtrj05j2tfn8ls7n3fhemg3sr.apps.googleusercontent.com
    clientSecret:
      group: ""
      kind: Secret
      name: my-app-client-secret
    provider:
      issuer: https://accounts.google.com
  targetRef:
    group: gateway.networking.k8s.io
    kind: Gateway
    name: eg
```
Then you'd apply it like
```
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: oidc-example
spec:
  oidc:
    configurations:
      enable:
        - name: oidc-config-default
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: backend-root
```
and for unprotected routes like:
```
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: oidc-example
spec:
  oidc:
    configurations:
      disable:
        - name: oidc-config-default
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: backend-unprotected
```

or vice versa. 

The `SecurityConfiguration` approach would provide more clarity and certainty as to how security configurations are applied -and also be more in line with the way filters actually work in envoy. 

Right now application is a bit ambiguous. For instance I tried initially working around this problem by applying an empty `SecurityFilter` at /unprotected thinking it would disable filters defined at / for /unprotected -but as of now an empty security filter appears to have no effect.

**Comment by zhaohuabing** [comment]
> @zhaohuabing What about creating an option like `passthrough: true` in the oidc config and utilizing the passthrough matcher?
It makes perfect sense to add a `passthrough` option in the OIDC configuration because some sub-paths of an HTTPRoute may not need authentication.

But we still need to expose the `redirect_uri` to the API if we want to solve the previous issue, right?

Let's say there're two applications "bookstore" and "petstore", and we would like to define a `SecurityPolicy` for each one to configure its OIDC aunthentication. For each application, there're some public pages that don't need authentication.

HTTPRoutes for "bookstore" and "petstore":
```yaml
kind: HTTPRoute
metadata:
  name: bookstore
spec:
  hostnames:
  - www.example.com
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: eg
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /bookstore
    backendRefs:
    - group: ""
      kind: Service
      name: bookstore
      port: 3000
  ```    
  
  yaml
  ```
  kind: HTTPRoute
metadata:
  name: petstore
spec:
  hostnames:
  - www.example.com
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: eg
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /petstore
    backendRefs:
    - group: ""
      kind: Service
      name: petstore
      port: 3000
  ```
  
  SecurityPolicies for "bookstore" and "petstore"
  ```yaml
  apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: bookstore
spec:
  oidc:
    redirect_uri: /bookstore/oauth2/callback # make sure the redirect_uri can be caught by HTTPRoute bookstore
    pass_through:  # exclude public pages
    - path:
        type: PathPrefix
        value: /petstore/pulic-page1
    - path:
        type: PathPrefix
        value: /petstore/pulic-page1    
    clientID: bookstore_client_id
    clientSecret:
      group: ""
      kind: Secret
      name: book-store-client-secret
    provider:
      issuer: https://oauth2.myoidcprovider.com
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: bookstore
```

```yaml
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: petstore
spec:
  oidc:
    redirect_uri: /bookstore/oauth2/callback # make sure the redirect_uri can be caught by HTTPRoute petstore
    pass_through:  # exclude public pages 
    - path:
        type: PathPrefix
        value: /petstore/pulic-page1
    - path:
        type: PathPrefix
        value: /petstore/pulic-page1    
    clientID: bookstore_client_id
    clientSecret:
      group: ""
      kind: Secret
      name: petstore-client-secret
    provider:
      issuer: https://oauth2.myoidcprovider.com
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: petstore
```

**Comment by apjoseph** [comment]
@zhaohuabing in your example, yes -and I think it's  good idea to be able to set logout/redirect_uri directly -but I was thinking something more along the lines of:

```
kind: HTTPRoute
metadata:
  name: example-combined
spec:
  hostnames:
  - www.example.com
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: eg
  rules:
    backendRefs:
     - group: ""
      kind: Service
      name: bookstore
      port: 3000
   - group: ""
      kind: Service
      name: petstore
      port: 3000
 ```
  
 ```
 apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
name: example-main
spec:
oidc:
  redirect_uri: /oauth2/callback
  logout_url: /logout 
   # exclude public pages
  paths:
      - type: PathPrefix
         value: /
        behavior: disable
      - type: PathPrefix  
         value: /petstore/private-page1
         behavior: enable
      - type: PathPrefix  
         value: /bookstore/private-page1
         behavior: enable
  clientID: combined_client_id
  clientSecret:
    group: ""
    kind: Secret
    name: combined-client-secret
  provider:
    issuer: https://oauth2.myoidcprovider.com
targetRef:
  group: gateway.networking.k8s.io
  kind: HTTPRoute
  name:  example-combined
```

# AND/OR
```
 apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
name: example-main
spec:
oidc:
  redirect_uri: /oauth2/callback
  logout_url: /logout 
   # exclude public pages
  paths:
      - type: PathPrefix
         value: /
        behavior: disable
      - type: PathPrefix  
         value: /petstore/private-page1
         behavior: enable
      - type: PathPrefix  
         value: /bookstore/private-page2
         behavior: enable
  clientID: combined_client_id
  clientSecret:
    group: ""
    kind: Secret
    name: combined-client-secret
  provider:
    issuer: https://oauth2.myoidcprovider.com
targetRef:
  group: gateway.networking.k8s.io
  kind: Gateway
  # sectionName would be optional 
  sectionName: my-listenter
  name:  eg
```
By attaching the SecurityPolicy to the gateway and/or listener, the user can avoid defining the unecessary `example-main` HTTPRoute since it would basically have the same effect as attaching to the gateway in this case.  

downside of either approach I laid out here is that you'd have to define all of the security routing at the top level. 

However if you have multiple services that make up one logical application hosted at www.example.com, you'd definitely not want to define callback urls for every different service -since it mean you have to edit your oauth2 provider's configuration every time you get a new backend. 

Imagine instead of `bookstore` and `petstore` you have `bookstore-next` and `bookstore-angular` where an aging angular app is being slowly being migrated to next.js -but both just serve different routes of the same bookstore site for instance. 

The `SecurityConfiguration` CRD in my last post would solve both problems -but obviously that would be a much bigger change to implement.

**Comment by zhaohuabing** [comment]
@apjoseph 

```yaml
kind: HTTPRoute
metadata:
  name: example-combined
spec:
  hostnames:
  - www.example.com
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: eg
  rules:
    backendRefs:
     - group: ""
      kind: Service
      name: bookstore
      port: 3000
   - group: ""
      kind: Service
      name: petstore
      port: 3000
```
This HTTPRoute routes request to both the `bookstore` and `petstore` equally? 

> Imagine instead of bookstore and petstore you have bookstore-next and bookstore-angular where an aging angular app is being slowly being migrated to next.js -but both just serve different routes of the same bookstore site for instance.

You can have one HTTPRoute with multiple rules, one rule for the old `bookstore`, and each rule for the new `bookstore-next` with different paths. The following is an example of the HTTPRoute CRD:
```yaml
kind: HTTPRoute
metadata:
  name: bookstore
spec:
  hostnames:
  - www.example.com
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: eg
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /bookstore
    backendRefs:
    - group: ""
      kind: Service
      name: bookstore
      port: 3000
    - path:
        type: PathPrefix
        value: /bookstore/upgrated-path1
    backendRefs:
    - group: ""
      kind: Service
      name: bookstore-next
      port: 3000
    - path:
        type: PathPrefix
        value: /bookstore/upgrated-path2
    backendRefs:
    - group: ""
      kind: Service
      name: bookstore-next
      port: 3000
```

SecurityPolicy

```yaml
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
name: bookstore
spec:
oidc:
  # make sure the redirect_uri can be caught by HTTPRoute bookstore
  redirect_uri: /bookstore/oauth2/callback 
   # exclude public pages
  pass_through: 
  - path:
      type: PathPrefix
      value: /petstore/pulic-page1
  - path:
      type: PathPrefix
      value: /petstore/pulic-page2    
  clientID: bookstore_client_id
  clientSecret:
    group: ""
    kind: Secret
    name: bookstore-client-secret
  provider:
    issuer: https://oauth2.myoidcprovider.com
targetRef:
  group: gateway.networking.k8s.io
  kind: HTTPRoute
  name: bookstore
```

**Comment by zhaohuabing** [comment]
@apjoseph 

```yaml
kind: HTTPRoute
metadata:
  name: example-combined
spec:
  hostnames:
  - www.example.com
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: eg
  rules:
    backendRefs:
     - group: ""
      kind: Service
      name: bookstore
      port: 3000
     - group: ""
      kind: Service
      name: petstore
      port: 3000
```
This HTTPRoute routes request to both the `bookstore` and `petstore` equally? 

> Imagine instead of bookstore and petstore you have bookstore-next and bookstore-angular where an aging angular app is being slowly being migrated to next.js -but both just serve different routes of the same bookstore site for instance.


The bookstore and boostore-next should be treated as the same application and have the same OIDC configuration.

You can have one HTTPRoute with multiple rules, one rule for the old `bookstore`, and each other rule for the new `bookstore-next` with different paths. The following is an example of the HTTPRoute:
```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: bookstore
spec:
  hostnames:
  - www.example.com
  parentRefs:
  - group: gateway.networking.k8s.io
    kind: Gateway
    name: eg
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /bookstore
    backendRefs:
    - name: bookstore
      port: 3000
  - matches:
    - path:
        type: PathPrefix
        value: /bookstore/upgrated-path1
    backendRefs:
    - name: bookstore-next
      port: 3000
  - matches:
    - path:
        type: PathPrefix
        value: /bookstore/upgrated-path2
    backendRefs:
    - name: bookstore-next
      port: 3000
```

SecurityPolicy

```yaml
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
name: bookstore
spec:
oidc:
  # make sure the redirect_uri can be caught by HTTPRoute bookstore
  redirect_uri: /bookstore/oauth2/callback 
   # exclude public pages
  pass_through: 
  - path:
      type: PathPrefix
      value: /bookstore/pulic-page1
  - path:
      type: PathPrefix
      value: /bookstore/pulic-page2    
  clientID: bookstore_client_id
  clientSecret:
    group: ""
    kind: Secret
    name: bookstore-client-secret
  provider:
    issuer: https://oauth2.myoidcprovider.com
targetRef:
  group: gateway.networking.k8s.io
  kind: HTTPRoute
  name: bookstore
```

**Comment by zirain** [comment]
@zhaohuabing fixed by #2409?

**Comment by arkodg** [comment]
@zhaohuabing can this issue be closed and any remaining sub task be tracked with a another GH issue ?
